### PR TITLE
[CPU] Support CPU W4A16 INT4 MoE

### PR DESCRIPTION
**Pull request by yuwenzho**
## Purpose

Support cpu AWQ, GPTQ & compressed-tensor W4A16 INT4 MoE for Intel CPUs.

## Test Plan
python -m pytest tests/kernels/moe/test_cpu_quant_fused_moe.py -v

## Test Result

**1. AWQ W4A16 INT4 MoE**
python examples/basic/offline_inference/generate.py --model QuixiAI/Qwen3-30B-A3B-AWQ --max-model-len 512 --enforce-eager
```
--------------------------------------------------
Prompt: 'Hello, my name is'
Generated text: ' James and I am a 23-year-old male. I have been experiencing'
--------------------------------------------------
Prompt: 'The president of the United States is'
Generated text: ' the head of state and head of government of the United States, and is the'
--------------------------------------------------
Prompt: 'The capital of France is'
Generated text: ' Paris. The capital of Germany is Berlin. The capital of the United States is'
--------------------------------------------------
Prompt: 'The future of AI is'
Generated text: ' a topic that has sparked intense debate and speculation, ranging from utopian visions of'
--------------------------------------------------
```

**2. GPTQ W4A16 INT4 MoE**
python examples/basic/offline_inference/generate.py --model Qwen/Qwen3-30B-A3B-GPTQ-Int4  --max-model-len 512 --enforce-eager
```
--------------------------------------------------
Prompt: 'Hello, my name is'
Generated text: ' James and I am a third-year student majoring in Computer Science at the University'
--------------------------------------------------
Prompt: 'The president of the United States is'
Generated text: ' the head of the executive branch of the federal government. He is the chief of'
--------------------------------------------------
Prompt: 'The capital of France is'
Generated text: ' Paris. The capital of Italy is Rome. The capital of Spain is Madrid.'
--------------------------------------------------
Prompt: 'The future of AI is'
Generated text: ' a topic that has sparked intense debate and speculation, particularly in the context of its'
--------------------------------------------------
```

**3.   Compressed-tensor W4A16 INT4 MoE**
python examples/basic/offline_inference/generate.py --model RedHatAI/Qwen3-30B-A3B-quantized.w4a16  --max-model-len 512 --enforce-eager
```
--------------------------------------------------
Prompt: 'Hello, my name is'
Generated text: ' James and I am a third-year medical student. I am applying for the '
--------------------------------------------------
Prompt: 'The president of the United States is'
Generated text: ' the head of state and head of government of the United States. He or she'
--------------------------------------------------
Prompt: 'The capital of France is'
Generated text: ' Paris. Paris is a city in Europe, and it is known for its art'
--------------------------------------------------
Prompt: 'The future of AI is'
Generated text: ' a topic that has sparked intense debate and speculation, ranging from utopian visions of'
--------------------------------------------------
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by yuwenzho** [comment]
@bigPYJ1151 please review

**Comment by yuwenzho** [comment]
@gemini-code-assist review

File: vllm/model_executor/layers/fused_moe/oracle/int_wna16.py [hunk 1]
```diff
@@ -857,6 +870,85 @@ def convert_to_wna16_moe_kernel_format(
...
+            w2_g_idx,
+            None,  # w13_g_idx_sort_indices (unused on CPU)
+            None,  # w2_g_idx_sort_indices (unused on CPU)
+            blocked_z13,
+            blocked_z2,
+            None,  # w13_input_global_scale
+            None,  # w2_input_global_scale
+            w13_bias,
+            w2_bias,
+        )
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/fused_moe/oracle/int_wna16.py:997]: Perhaps the process procedure can be moved to ```cpu_moe.py```, likes other backends' implementation,

File: vllm/model_executor/layers/fused_moe/oracle/int_wna16.py [hunk 2]
```diff
@@ -857,6 +870,85 @@ def convert_to_wna16_moe_kernel_format(
...
+        return (
+            blocked_w13,
+            blocked_w2,
+            blocked_s13,
+            blocked_s2,
+            w13_g_idx,
+            w2_g_idx,
+            None,  # w13_g_idx_sort_indices (unused on CPU)
+            None,  # w2_g_idx_sort_indices (unused on CPU)
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/fused_moe/oracle/int_wna16.py]: Should raise error for GPTQ with ```desc_act=True```

File: vllm/model_executor/layers/fused_moe/oracle/int_wna16.py [hunk 3]
```diff
@@ -857,6 +870,85 @@ def convert_to_wna16_moe_kernel_format(
...
+            blocked_z13,
+            blocked_z2,
+            None,  # w13_input_global_scale
+            None,  # w2_input_global_scale
+            w13_bias,
+            w2_bias,
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/fused_moe/oracle/int_wna16.py:996]: bias also needs to be converted to ```float32```, if they are not ```None```.

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_wna16.py [hunk 4]
```diff
@@ -51,6 +60,19 @@ def __init__(
...
+        self.moe_kernel: mk.FusedMoEKernel | None = None
+        if current_platform.is_cpu():
+            if self.num_bits != 4:
+                raise NotImplementedError(
+                    "CPU WNA16 MoE currently supports W4A16 only."
+                )
+            self.wna16_moe_backend, self.experts_cls = select_wna16_moe_backend(
+                moe,
+                kInt4Static,
+            )
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_wna16.py]: Looks like we should use ```compressed_tensors_moe_wna16_marlin.py```, it includes new MoE backend abstraction to select kernels.
I feel this implementation is a bit legacy and might be deprecated in the future.

File: vllm/model_executor/layers/quantization/auto_gptq.py [hunk 5]
```diff
@@ -640,8 +640,13 @@ def create_weights(
         layer.register_parameter("w2_g_idx_sort_indices", w2_g_idx_sort_indices)
         set_weight_attrs(w2_g_idx_sort_indices, extra_weight_attrs)
 
-        device = layer.w13_qweight.device
-        layer.workspace = marlin_make_workspace_new(device, 4)
+        from vllm.model_executor.layers.fused_moe.oracle.int_wna16 import (
+            WNA16MoEBackend,
+        )
+
+        if self.wna16_moe_backend != WNA16MoEBackend.CPU:
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/auto_gptq.py]: In current MoE backend abstraction, ```FusedMoEExpertsModular``` requires workspace allocation and ```FusedMoEExpertsMonolithic``` doesn't.

Based on this we can avoid add the CPU specific check.

File: vllm/model_executor/layers/fused_moe/experts/cpu_moe.py [hunk 6]
```diff
@@ -318,3 +324,201 @@ def apply(
...
+        global_num_experts: int,
+        expert_map: torch.Tensor | None,
+        a1q_scale: torch.Tensor | None,
+        apply_router_weight_on_input: bool,
+        # grouped topk + fused topk bias parameters
+        num_expert_group: int | None = None,
+        e_score_correction_bias: torch.Tensor | None = None,
+        routed_scaling_factor: float | None = None,
+        topk_group: int | None = None,
+    ) -> torch.Tensor:
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/fused_moe/experts/cpu_moe.py:484]: Better to check ```apply_router_weight_on_input``` here to avoid silent accuracy bug.
In most models it is ```False```, only ```Llama4``` uses it.

**Review by bigPYJ1151 — COMMENTED** [review]
Thanks for the great work @yuwenzho :) I have some comments, please check them.

And please also add related cases to ```test_cpu_wna16.py``` for E2E check.

File: vllm/model_executor/layers/fused_moe/oracle/int_wna16.py (hunk 1 above)
Comment (yuwenzho) [reply vllm/model_executor/layers/fused_moe/oracle/int_wna16.py:997]: I extracted the inline CPU processing block into a dedicated _process_weights_cpu() function. I keep it in int_wna16.py to align with _process_weights_xpu(), _process_weights_marlin() rather than moving it to cpu_moe.py

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_wna16_marlin.py [hunk 7]
```diff
@@ -414,8 +415,9 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         replace_parameter(layer, "w13_weight_scale", w13_scales)
         replace_parameter(layer, "w2_weight_scale", w2_scales)
 
-        if not self.symmetric:
+        if w13_qzeros is not None:
```
Comment (bigPYJ1151) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe/compressed_tensors_moe_wna16_marlin.py:418]: Hmmm, I'm not sure whether this replacement is safe because different backends may have different behaviors. I would suggest to keep it as original.

File: vllm/model_executor/layers/fused_moe/oracle/int_wna16.py
Comment (bigPYJ1151) [inline vllm/model_executor/layers/fused_moe/oracle/int_wna16.py:280]: New arguments here, need to update ```CPUExpertsInt4```

**Review by bigPYJ1151 — APPROVED** [review]

**Comment by eugr** [comment]
@yuwenzho @bigPYJ1151 - This PR broke Autoround quants on NVIDIA GPUs, specifically this change:

```diff
diff --git a/vllm/model_executor/layers/quantization/auto_gptq.py b/vllm/model_executor/layers/quantization/auto_gptq.py
@@ -750,12 +758,8 @@ def get_fused_moe_quant_config(self, layer: RoutedExperts) -> FusedMoEQuantConfi
             w2_scale=layer.w2_scales,
             weight_bits=self.quant_config.weight_bits,
             group_size=self.quant_config.group_size,
-            w1_zp=getattr(layer, "w13_qzeros", None)
-            if not self.quant_config.is_sym
-            else None,
-            w2_zp=getattr(layer, "w2_qzeros", None)
-            if not self.quant_config.is_sym
-            else None,
+            w1_zp=getattr(layer, "w13_qzeros", None),
+            w2_zp=getattr(layer, "w2_qzeros", None),
             w1_bias=getattr(layer, "w13_bias", None),
             w2_bias=getattr(layer, "w2_bias", None),
         )
```

After reverting this hunk, everything works again.

@mgoin - FYI.

**Comment by yuwenzho** [comment]
> @yuwenzho @bigPYJ1151 - This PR broke Autoround quants on NVIDIA GPUs, specifically this change:
> 
> ```diff
> diff --git a/vllm/model_executor/layers/quantization/auto_gptq.py b/vllm/model_executor/layers/quantization/auto_gptq.py
> @@ -750,12 +758,8 @@ def get_fused_moe_quant_config(self, layer: RoutedExperts) -> FusedMoEQuantConfi
>              w2_scale=layer.w2_scales,
>              weight_bits=self.quant_config.weight_bits,
>              group_size=self.quant_config.group_size,
> -            w1_zp=getattr(layer, "w13_qzeros", None)
> -            if not self.quant_config.is_sym
> -            else None,
> -            w2_zp=getattr(layer, "w2_qzeros", None)
> -            if not self.quant_config.is_sym
> -            else None,
> +            w1_zp=getattr(layer, "w13_qzeros", None),
> +            w2_zp=getattr(layer, "w2_qzeros", None),
>              w1_bias=getattr(layer, "w13_bias", None),
>              w2_bias=getattr(layer, "w2_bias", None),
>          )
> ```
> 
> After reverting this hunk, everything works again.
> 
> @mgoin - FYI.

@eugr Thanks for the report. A fix PR #45656 has been submitted that restores the guard with an explicit CPU exception, so symmetric GPU quants get None (original behavior) while symmetric CPU gets the synthesized zp it needs. Please help verify and review.
